### PR TITLE
GUI: Added configurable keyboard mappings to the project

### PR DESCRIPTION
**Pull request by flplv**
This PR adds the support to configure the keybindings using the settings dialog. 

It also rollback a recent commit that bound mouse button to the touch pad of the PS4 controller, now just go the settings screen and bind a key to the pad (@3kinox, please let me know if you accept this change). 

Next feature I will develop is an autoclick functionality. Let me know where do you believe that should be, at GUI or chiaki library?

I suggest you review this PR commit by commit, so you can avoid reading some of the formatting fixes in the settings.cpp file.

### CODE REVIEW

**Comment by juicypop** [comment]
@flplv would this solve my issue #169?

**Comment by flplv** [comment]
Yes, but analog channels are treat as a digital one. Depending on the game it might be annoying.

File: gui/include/settings.h [hunk 1]
```diff
@@ -24,11 +24,22 @@
...
+	// must not overlap with ChiakiControllerButton and ChiakiControllerAnalogButton
+	CHIAKI_CONTROLLER_ANALOG_STICK_LEFT_X_UP = (1 << 18),
+	CHIAKI_CONTROLLER_ANALOG_STICK_LEFT_X_DOWN = (1 << 19),
+	CHIAKI_CONTROLLER_ANALOG_STICK_LEFT_Y_UP = (1 << 20),
+	CHIAKI_CONTROLLER_ANALOG_STICK_LEFT_Y_DOWN = (1 << 21),
+	CHIAKI_CONTROLLER_ANALOG_STICK_RIGHT_X_UP = (1 << 22),
+	CHIAKI_CONTROLLER_ANALOG_STICK_RIGHT_X_DOWN = (1 << 23),
+	CHIAKI_CONTROLLER_ANALOG_STICK_RIGHT_Y_UP = (1 << 24),
+	CHIAKI_CONTROLLER_ANALOG_STICK_RIGHT_Y_DOWN = (1 << 25),
+} SettingsExtendedAnalogStickEnum;
```
Comment (thestr4ng3r) [inline gui/include/settings.h]: I think it would be better to differentiate these a bit more from the original definitions. I would do it like this (no typedef needed in C++):
```c++
enum class ControllerButtonExt {
	ANALOG_STICK_LEFT_X_UP = (1 << 18),
	...
};
```
Then later use it like `static_cast<int>(ControllerButtonExt::ANALOG_STICK_LEFT_X_UP)`

File: gui/include/settings.h [hunk 2]
```diff
@@ -24,11 +24,22 @@
...
+
+class Settings : public QObject {
 	Q_OBJECT
 
-	private:
+ 	private:
```
Comment (thestr4ng3r) [inline gui/include/settings.h]: Unnecessary leading space

File: gui/include/settings.h [hunk 3]
```diff
@@ -41,14 +52,22 @@ class Settings : public QObject
 		void LoadManualHosts();
 		void SaveManualHosts();
 
-	public:
+ 	public:
```
Comment (thestr4ng3r) [inline gui/include/settings.h]: leading space

File: gui/include/settings.h [hunk 4]
```diff
@@ -41,14 +52,22 @@ class Settings : public QObject
...
 		explicit Settings(QObject *parent = nullptr);
 
-		bool GetDiscoveryEnabled() const		{ return settings.value("settings/auto_discovery", true).toBool(); }
-		void SetDiscoveryEnabled(bool enabled)	{ settings.setValue("settings/auto_discovery", enabled); }
-
-		bool GetLogVerbose() const 				{ return settings.value("settings/log_verbose", false).toBool(); }
-		void SetLogVerbose(bool enabled)		{ settings.setValue("settings/log_verbose", enabled); }
+		bool GetDiscoveryEnabled() const {
```
Comment (thestr4ng3r) [inline gui/include/settings.h]: Please keep the one-line formatting for these

File: gui/include/settingskeycapturedialog.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#include <QDialog>
+
+class SettingsKeyCaptureDialog : public QDialog {
+	Q_OBJECT
+
+	signals:
+		void KeyCaptured(Qt::Key);
+
+ 	protected:
+		void keyPressEvent(QKeyEvent *event) override;
```
Comment (thestr4ng3r) [inline gui/include/settingskeycapturedialog.h]: This isn't used, so can be removed.

File: gui/include/settingskeycapturedialog.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+		void keyReleaseEvent(QKeyEvent *event) override;
+
+ 	public:
+		SettingsKeyCaptureDialog(QWidget *parent = nullptr);
```
Comment (thestr4ng3r) [inline gui/include/settingskeycapturedialog.h]:
```suggestion
		explicit SettingsKeyCaptureDialog(QWidget *parent = nullptr);
```

File: gui/include/streamsession.h [hunk 7]
```diff
@@ -47,6 +48,7 @@ class ChiakiException: public Exception
 
 struct StreamSessionConnectInfo
 {
+ 	Settings * settings;
```
Comment (thestr4ng3r) [inline gui/include/streamsession.h]: This struct shouldn't depend on the actual `Settings` class. Better add a member of type `QMap<Qt::Key, int>` and extract the value in the `StreamSessionConnectInfo(Settings *settings, QString host, QByteArray regist_key, QByteArray morning)` constructor.

File: gui/src/settingskeycapturedialog.cpp [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	button->setAutoDefault(false);
+	connect(button, &QPushButton::clicked, this, &QDialog::reject);
+}
+
+void SettingsKeyCaptureDialog::keyPressEvent(QKeyEvent* event) {
+}
+
+void SettingsKeyCaptureDialog::keyReleaseEvent(QKeyEvent* event) {
+	KeyCaptured(Qt::Key(event->key()));
+	reject();
```
Comment (thestr4ng3r) [inline gui/src/settingskeycapturedialog.cpp]:
```suggestion
	accept();
```
Doesn't change the behavior in this case, but accept makes more sense.

File: gui/src/streamsession.cpp [hunk 9]
```diff
@@ -141,66 +136,55 @@ void StreamSession::SetLoginPIN(const QString &pin)
...
+		return;
+
+	if (event->isAutoRepeat())
+		return;
+
+	int button = key_map[Qt::Key(event->key())];
+	bool press_event = event->type() == QEvent::Type::KeyPress;
+
+	if (press_event)
+		keyboard_state.buttons |= button;
```
Comment (thestr4ng3r) [inline gui/src/streamsession.cpp]: `keyboard_state.buttons` should not be touched for the extended buttons. I think this whole `if(press_event)` could be moved into the `default` case of the switch below.

File: lib/include/chiaki/controller.h [hunk 10]
```diff
@@ -56,6 +56,7 @@ typedef enum chiaki_controller_analog_button_t
 	CHIAKI_CONTROLLER_ANALOG_BUTTON_R2 = (1 << 17)
 } ChiakiControllerAnalogButton;
 
+
```
Comment (thestr4ng3r) [inline lib/include/chiaki/controller.h]: Unnecessary change

**Comment by thestr4ng3r** [comment]
The reason for the mouse click for touchpad button was that Linux by default handles the Dualshock's touchpad as a mouse-like touchpad so if you use the actual controller you could use the real touchpad button. Mapping it to a keyboard key wouldn't help here, so please keep it for now.

What exactly do you mean by "autoclick"? Something like a turbo button that automatically presses a button in a fast interval?

**Comment by flplv** [comment]
Ah got it, just paired my PS4 controller to the PC and saw the issue.

Where do you keep user manual? in the README? I believe a section speaking how the Desktop client can be used could be useful, specifically to point out that PS4 can be paired and or keyboard can be mapped.

Autoclick is a feature that I particularly use, and that was what drove me to this project. It is a timed automatic clicker for AFK gaming, it just periodically click a given button or sequence of buttons. It is useful for games like Minecraft. In Minecraft we use for AFK Fishing to level up and collect enchantment books. If you find it too specialized to the project, that is fine, I can keep it on my fork. But If you want the feature in the project, I'd add another entry in the settings menu where the user can configure the timer and the button to auto click, then the internals of the autoclick could go either in the Gui or in the shared library. 

I will make some commits to address your inputs and I'll let you know.

File: gui/src/streamsession.cpp (hunk 9 above)
Comment (flplv) [reply gui/src/streamsession.cpp]: yes, you caught a bug, thanks.

**Comment by 3kinox** [comment]
@flplv For the setting + touchpad fix, one way is simply to give the option to use any input types (key, controller button, mouse, or whatever) for binding. Not sure if you can do it straight in QT or not. Otherwise, a cleaner and more versatile fix could be to use SDL or evdev for all inputs, they should also give you control over analog vs numeric sticks/ R2-L2 (and rumble).

File: gui/include/streamsession.h (hunk 7 above)
Comment (flplv) [reply gui/include/streamsession.h]: How do you feel about passing `Settings *` to `StreamWindow` constructor, in order to let it connect to a signal in `Setttings` that would be called `KeyboardMapChanged`? That would enable the user to change the mapping without having to reopen the stream.

Comment (flplv) [reply gui/include/streamsession.h]: or I can do what you say, but then at the same time connect the slot externally to StreamWindow. I think that would be better.

Comment (flplv) [reply gui/include/streamsession.h]: humm.. I realized that the `StreamWindow` is consuming all keypress, and not forwarding to other windows, so changing the key mapping while stream is on is not possible. Not sure if want to change this behavior.

**Comment by flplv** [comment]
> @flplv For the setting + touchpad fix, one way is simply to give the option to use any input types (key, controller button, mouse, or whatever) for binding. Not sure if you can do it straight in QT or not. Otherwise, a cleaner and more versatile fix could be to use SDL or evdev for all inputs, they should also give you control over analog vs numeric sticks/ R2-L2 (and rumble).

That is a good idea. I will keep in mind for the next improve.

**Comment by flplv** [comment]
@thestr4ng3r I believe I addressed all your comments.

**Comment by thestr4ng3r** [comment]
> Where do you keep user manual? in the README?

Yes, for now there is only the readme, but more extensive documentation is planned.

> Autoclick is a feature that I particularly use, and that was what drove me to this project. It is a timed automatic clicker for AFK gaming, it just periodically click a given button or sequence of buttons. It is useful for games like Minecraft. In Minecraft we use for AFK Fishing to level up and collect enchantment books. If you find it too specialized to the project, that is fine, I can keep it on my fork. But If you want the feature in the project, I'd add another entry in the settings menu where the user can configure the timer and the button to auto click, then the internals of the autoclick could go either in the Gui or in the shared library.

I see, well I'm not entirely sure I want that as part of the main project. But you can open an issue about it so we can discuss it in detail there.

**Comment by thestr4ng3r** [comment]
Thank you very much!
If you want to continue on this, a button to reset the key bindings to the default would be nice.
